### ISSUE TITLE
compact option doesn't print comments

### ISSUE THREAD

**Issue by amasad** [babel/babel#1810]
If compact is true then comments are not printed. Is this by design?

**Comment by loganfsmyth** [comment]
Yup: https://github.com/babel/babel/blob/fce977f1d73390eab0922e1cc7efa5cd91fba23b/src/babel/generation/index.js#L310

**Comment by amasad** [comment]
Do you know why?

**Comment by loganfsmyth** [comment]
I'm sure @sebmck can give an official answer, but if I had to guess, it's simply that compact exists to keep generation fast and if your code is already all compressed onto a single line, comments probably aren't that useful.

Do you have a specific use-case that you are hitting and having trouble with?

**Comment by amasad** [comment]
Fair enough. We have a docblock that has some semantic meaning that we need to print. We can't use the pretty printer because it's very slow, so I wanted to preserve the comment without pretty printing.

Another idea is that we parse and prepend the docblock on our end.

**Comment by loganfsmyth** [comment]
I guess that's a question for @sebmck. Perhaps Babel could preserve `/*! */` style comments to allow this use-case, his call.

**Comment by amasad** [comment]
cc @DmitrySoshnikov

**Comment by sebmck** [comment]
Yeah, it's probably worth having `/*!` comments be preserved unless `comments` has explicitly been set to `false`. Do the comments you want to retain follow that format?

**Comment by amasad** [comment]
Nope there of the following format at the very top of the file:

```
/**
 * @providesModule blah
 * @flow
 */
``

**Comment by sebmck** [comment]
Could either keep comments that contained `@` (can't think of any immediate problems) or `comments` could be overloaded to take a regex that matches the comments content to determine if it should be included.

**Comment by amasad** [comment]
Maybe keeping `@` comments is generally useful since it's the convention.

**Comment by loganfsmyth** [comment]
If we're making this an option, I'm pro-regex since blindly accepting anything with `@` is going to dump pretty much all the comments if your code uses jsdoc.

**Comment by DmitrySoshnikov** [comment]
Or maybe we can make them non mutual exclusive: if users would like to get the comments via the explicit `comments: true` (even in the `compact` mode), why limit them? Then the users may handled this value themselves from the outside, via the same regexp, or even simple (as long as they have pre-parsed dockblock, as in our case):

```javascript
comments: docblock.indexOf('@providesModule') !== -1
compact: true,
```

**Comment by loganfsmyth** [comment]
Mind if I ask about your usecase a bit more? I'm still not clear on why `compact: true` is something you are using here. What is making this file so huge? Perhaps you can move Babel earlier in whatever pipeline you are using?

**Comment by DmitrySoshnikov** [comment]
We don't need pretty printed output, so we use `compact` (in general, not specifically to some files; although, we also have some huge 3rd-party files which we don't control, but still need to run the local plugin transformers on them).

Regardless, I think if a user decides to get the comments in the output, we don't need to limit them, even in the compact mode (since currently it's confusing).

We have a workaround by manually appending the main dockblock after transformation, although seems it's better to provide the user ability to control whether they need the comments or not (especially if this option already exists).

**Comment by DmitrySoshnikov** [comment]
As for the printer in general, we've talked about this previously, in the future it'd be great not to reprint the things if they hasn't been changed. Although, it'll require some harder analysis of the "changed" state :smile: , so this is a separate topic for another issue. For now, Babel's pretty printer dies on a file containing 18000 lines. The fun thing is: the file doesn't even contain ES6 features, so it parsers, and reprints the whole program, for nothing. Since we need to apply custom plugins, we still run the transform, but this is just a small part of those 18000 lines. Compat printer seems is able to handle it though yet, but removes the comments which we need.

**Comment by DmitrySoshnikov** [comment]
> We have a workaround by manually appending the main dockblock after transformation

Actually, it doesn't work, since breaks line numbers. We have `retainLines` and `compact` combined, and still need to preserve main docblock (since further, not Babel, transforms depend on it).

For the code:

```js
/**
 * @providesModule Foo
 *
 * This is
 * a long
 * docblock
 * several
 * lines
 * yeah
 */

var id = (x) => x;
```

After manually prepending the output with the docblock, we got:

```js
/**
 * @providesModule Foo
 *
 * This is
 * a long
 * docblock
 * several
 * lines
 * yeah
 */
var










id=function(x){return x;};
```

Since `retainLines` actually retains them, it replaces those comments with whitespace. We could try to substring the output to the line numbers in the docblock, but there is no reliable way to cut them in correct position.

**Comment by sebmck** [comment]
Should be fixed via e50babe916f99ee36d0a2fa602195596985dd9f2. I've made two changes:

 - An optional `shouldPrintComment` option that allows you to specify a callback that controls whether or not specific comments are printed.
 - Made comments be printed in `compact` mode.

**Comment by sebmck** [comment]
Released as of 5.7.0, thanks! :blush:
